libraries/standard/mqtt/src/mqtt.c

```diff
@@ -130,6 +234,447 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+            break;
+        default:
+            /* Nothing to send. */
+            break;
+    }
+
+    if( packetTypeByte != 0U )
+    {
+        status = MQTT_SerializeAck( &( pContext->networkBuffer ),
+                                    packetTypeByte,
+                                    packetId );
+        if( status == MQTTSuccess )
+        {
+            bytesSent = sendPacket( pContext,
+                                    pContext->networkBuffer.pBuffer,
+                                    MQTT_PUBLISH_ACK_PACKET_SIZE );
+        }
+
+        if( bytesSent == ( int32_t ) MQTT_PUBLISH_ACK_PACKET_SIZE )
+        {
+            pContext->controlPacketSent = true;
+            newState = MQTT_UpdateStateAck( pContext,
+                                            packetId,
+                                            packetType,
+                                            MQTT_SEND );
+
+            if( newState == MQTTStateNull )
+            {
+                LogErrorWithArgs( "Failed to update state of publish %u.",
+                                  packetId );
+                status = MQTTIllegalState;
+            }
+        }
+        else
+        {
+            LogErrorWithArgs( "Failed to send ACK packet: PacketType=%02x, "
+                              "SentBytes=%d, "
+                              "PacketSize=%u",
+                              packetTypeByte,
+                              bytesSent,
+                              MQTT_PUBLISH_ACK_PACKET_SIZE );
+            status = MQTTSendFailed;
+        }
+    }
+
+    if( ( status == MQTTSuccess ) && ( newState != MQTTStateNull ) )
+    {
+        *pPublishState = newState;
+    }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleKeepAlive( MQTTContext_t * const pContext )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint32_t now = 0U, keepAliveMs = 0U;
+
+    assert( pContext != NULL );
+    now = pContext->callbacks.getTime();
+    keepAliveMs = 1000U * ( uint32_t ) pContext->keepAliveIntervalSec;
+
+    /* If keep alive interval is 0, it is disabled. */
+    if( ( keepAliveMs != 0U ) && 
+        ( calculateElapsedTime( now, pContext->lastPacketTime ) > keepAliveMs ) )
+    {
+        if( pContext->waitingForPingResp )
+        {
+            /* Has time expired? */
+            if( calculateElapsedTime( now, pContext->pingReqSendTimeMs ) > 
+                pContext->pingRespTimeoutMs )
+            {
+                status = MQTTKeepAliveTimeout;
+            }
+        }
+        else
+        {
+            status = MQTT_Ping( pContext );
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleIncomingPublish( MQTTContext_t * const pContext,
+                                           MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
```

Comment:
In general, I usually prefer to see values initialized with failure values rather than success ones.  In this case, you don't need an initializer at all since initialization is guaranteed at line 570.

Reply:
Changed